tools/c7n_mailer/c7n_mailer/splunk_delivery.py

```diff
@@ -0,0 +1,269 @@
+            except JsonPointerException:
+                continue
+            msg = JsonPatch([{'op': 'remove', 'path': path}]).apply(msg)
+        return msg
+
+    def deliver_splunk_messages(self, payloads):
+        """
+        Deliver log messages to Splunk,
+
+        :param payloads: list of payload dicts to send to Splunk
+        :type payloads: list
+        """
+        failed = 0
+        for payload in payloads:
+            if not self._try_send(payload):
+                failed += 1
+        if failed != 0:
+            raise RuntimeError(
+                'ERROR: {failed} of {count} Splunk HEC messages '
+                'failed to deliver.'.format(
+                    failed=failed, count=len(payloads)
+                )
+            )
+
+    def _try_send(self, payload):
+        """
+        Retry sending payload to splunk via ``_send_splunk`` up to
+        ``config["splunk_max_attempts"]`` times, sleeping a random amount
+        of time between 1 and 4 seconds between each try.
+
+        :param payload: the payload dict to send to Splunk as event data.
+        :type payload: dict
+        :return: True if sent successfully, False otherwise
+        :rtype: bool
+        """
+        max_attempts = self.config.get('splunk_max_attempts', 4)
+        maxlen = self.config.get('splunk_hec_max_length', None)
+        p = json.dumps(payload)
+        if maxlen is not None and len(p) > maxlen:
+            # This is in place for Splunk installations that are configured
+            # with a short maximum message length (i.e. 10,000 characters).
+            self.logger.error(
+                'ERROR: Sending %d characters to Splunk HEC; line length '
+                'limit is %d characters. Data will be truncated: %s',
+                len(p), maxlen, p
+            )
+        for i in range(0, max_attempts):
+            try:
+                self._send_splunk(p)
+                if maxlen is not None and len(p) > maxlen:
+                    # never retry if we're over configured max length
+                    return False
+                return True  # if no exception, just return
+            except Exception:
+                sleep_sec = uniform(1, 4)  # random float 1 to 4
+                self.logger.warning(
+                    'Caught exception sending to Splunk; '
+                    'retry in %s seconds', sleep_sec
+                )
+                sleep(sleep_sec)
+        self.logger.error(
+            'ERROR - Could not POST to Splunk after %d tries.', max_attempts
+        )
+        return False
+
+    def _send_splunk(self, payload):
+        """
+        Perform the actual data send to Splunk HEC for one log entry
+
+        :param payload: the JSON-encoded payload to send to Splunk as
+          event data
+        :type payload: str
+        """
+        url = self.config['splunk_hec_url']
+        self.logger.debug('Send to Splunk (%s): %s', url, payload)
```

Comment:
can we ditch the payload logging here as overly verbose, or move it behind an explicit config flag if its needed.

Reply:
Right now debug-level logging in mailer is [guarded by the mailer "debug" config option](https://github.com/cloud-custodian/cloud-custodian/blob/11bee9d7a709faebd6245e0e899d45b3110e3bdf/tools/c7n_mailer/c7n_mailer/sqs_queue_processor.py#L84-L86) and runs [at INFO level normally](https://github.com/cloud-custodian/cloud-custodian/blob/11bee9d7a709faebd6245e0e899d45b3110e3bdf/tools/c7n_mailer/c7n_mailer/deploy.py#L34). Do you still think the payload logging needs to be removed or handled specially? I'd assumed that if anyone specifically enables the debug setting, they want the lowest-level debug messages possible...

Reply:
ah ic, cool, if we're defaulting to info, i'm fine with verbose debug then.